### ISSUE TITLE
Unable to mount /sys/fs/cgroup

### ISSUE THREAD

**Issue by lmickh** [chef-cookbooks/docker#556]
I'm trying to run the ecs-agent container via this cookbook.  Normally it would include the following in a docker run command:

```
--volume=/sys/fs/cgroup:/sys/fs/cgroup:ro
```

I've tried adding it via both binds and volume at the container resource.  Either way I get a error starting the container like so:

```
System error: mkdir /var/lib/docker/aufs/mnt/<container-ID>/sys/fs/cgroup:: read-only file system
```

Works fine if I execute it manually with 'docker run'.  Not sure what would be different between the two, but the cookbook doesn't seem to function the same way as the 'docker run' command.  That or I've messed up the resource somehow and just don't see it. Below is the docker run command that works and the chef resource example that fails to start.

```
docker run --name ecs-agent --detach=true --restart=on-failure:10 \
--volume=/var/run/docker.sock:/var/run/docker.sock \
--volume=/var/log/ecs/:/log \
--volume=/var/lib/ecs/data:/data \
--volume=/sys/fs/cgroup:/sys/fs/cgroup:ro \
--volume=/var/run/docker/execdriver/native:/var/lib/docker/execdriver/native:ro \
--publish=127.0.0.1:51678:51678 \
--env=ECS_LOGFILE=/log/ecs-agent.log \
--env=ECS_LOGLEVEL=info \
--env=ECS_DATADIR=/data \
--env=ECS_CLUSTER=default \
amazon/amazon-ecs-agent:v1.6.0
```

```
docker_container 'ecs-agent' do
  repo 'amazon/amazon-ecs-agent'
  env [ "ECS_LOGFILE=/log/ecs-agent.log",
        "ECS_LOGLEVEL=#{node['ecs']['log_level']}",
        "ECS_DATADIR=/data",
        "ECS_CLUSTER=#{node['ecs']['cluster']}"
  ]
 volume [ "#{node['ecs']['log_dir']}:/log",
           "/var/run/docker.sock:/var/run/docker.sock",
           "#{node['ecs']['data_dir']}:/data",
           "/sys/fs/cgroup:/sys/fs/cgroup:ro",
           "/var/run/docker/execdriver/native:/var/lib/docker/execdriver/native:ro"
  ]
  port '127.0.0.1:51678:51678'
  detach true
  tag node['ecs']['tag']
  restart_policy 'on-failure'
  restart_maximum_retry_count 10
end
```

EDIT:  Using docker 1.9.0 and version 2.2.6 of this cookbook.

**Comment by chasebolt** [comment]
Change volume to bind and gist the stack trace. Also what cookbook and docker version?

**Comment by lmickh** [comment]
https://gist.github.com/lmickh/19548b4c9ebfad77f0d2

Updated the original with version numbers.

**Comment by chasebolt** [comment]
Can you update to the latest version of this cookbook and see if it has been resolved?

**Comment by someara** [comment]
Currently, the "volumes" property needs to be "binds" until #493 is resolved

**Comment by lmickh** [comment]
That gist was produced using the binds option.  The cookbook version was a typo.  I was using 2.2.6.  I've upgraded to 2.2.8 and the behavior is the same.

**Comment by someara** [comment]
what platform are you using?

**Comment by lmickh** [comment]
Debian 8

**Comment by lmickh** [comment]
Decided to packet dump the api call of the cli and compare it to the chef/ruby code.

This is the docker run cli:

```
{"Hostname":"","Domainname":"","User":"","AttachStdin":false,"AttachStdout":false,"AttachStderr":false,"ExposedPorts":{"51678/tcp":{}},"Tty":false,"OpenStdin":false,"StdinOnce":false,"Env":["ECS_LOGFILE=/log/ecs-agent.log","ECS_LOGLEVEL=info","ECS_DATADIR=/data","ECS_CLUSTER=default"],"Cmd":null,"Image":"amazon/amazon-ecs-agent:v1.6.0","Volumes":{},"WorkingDir":"","Entrypoint":null,"OnBuild":null,"Labels":{},"StopSignal":"SIGTERM","HostConfig":{"Binds":["/var/run/docker.sock:/var/run/docker.sock","/var/log/ecs/:/log","/var/lib/ecs/data:/data","/sys/fs/cgroup:/sys/fs/cgroup:ro","/var/run/docker/execdriver/native:/var/lib/docker/execdriver/native:ro"],"ContainerIDFile":"","LxcConf":[],"Memory":0,"MemoryReservation":0,"MemorySwap":0,"KernelMemory":0,"CpuShares":0,"CpuPeriod":0,"CpusetCpus":"","CpusetMems":"","CpuQuota":0,"BlkioWeight":0,"OomKillDisable":false,"MemorySwappiness":-1,"Privileged":false,"PortBindings":{"51678/tcp":[{"HostIp":"127.0.0.1","HostPort":"51678"}]},"Links":null,"PublishAllPorts":false,"Dns":null,"DnsOptions":null,"DnsSearch":null,"ExtraHosts":null,"VolumesFrom":null,"Devices":[],"NetworkMode":"default","IpcMode":"","PidMode":"","UTSMode":"","CapAdd":null,"CapDrop":null,"GroupAdd":null,"RestartPolicy":{"Name":"on-failure","MaximumRetryCount":10},"SecurityOpt":null,"ReadonlyRootfs":false,"Ulimits":null,"LogConfig":{"Type":"","Config":{}},"CgroupParent":"","ConsoleSize":[0,0],"VolumeDriver":""}}
```


And this is the Chef/Ruby code:

```
{"Image":"amazon/amazon-ecs-agent:v1.6.0","Labels":{},"Cmd":null,"AttachStderr":true,"AttachStdin":false,"AttachStdout":true,"Domainname":"","Entrypoint":["/agent"],"Env":["ECS_LOGFILE=/log/ecs-agent.log","ECS_LOGLEVEL=info","ECS_DATADIR=/data","ECS_CLUSTER=default","ECS_RESERVED_MEMORY=1024"],"ExposedPorts":{"51678/tcp":{}},"Hostname":"4f85c003f470","MacAddress":"","NetworkDisabled":false,"OpenStdin":false,"StdinOnce":false,"Tty":false,"User":"","Volumes":{"/sys/fs/cgroup:/sys/fs/cgroup:ro":{},"/var/lib/ecs/data:/data":{},"/var/log/ecs:/log":{},"/var/run/docker.sock:/var/run/docker.sock":{},"/var/run/docker/execdriver/native:/var/lib/docker/execdriver/native:ro":{}},"WorkingDir":"","HostConfig":{"Binds":["/var/log/ecs:/log","/var/run/docker.sock:/var/run/docker.sock","/var/lib/ecs/data:/data","/sys/fs/cgroup:/sys/fs/cgroup","/var/run/docker/execdriver/native:/var/lib/docker/execdriver/native"],"CapAdd":null,"CapDrop":null,"CgroupParent":"","CpuShares":0,"CpusetCpus":"","Devices":null,"Dns":null,"DnsSearch":null,"ExtraHosts":null,"Links":null,"LogConfig":null,"Memory":0,"MemorySwap":-1,"NetworkMode":"default","Privileged":false,"PortBindings":{"51678/tcp":[{"HostIp":"127.0.0.1","HostPort":"51678"}]},"PublishAllPorts":false,"RestartPolicy":{"Name":"on-failure","MaximumRetryCount":10},"Ulimits":null,"VolumesFrom":null}}
```

Looks like the code makes the api call as a volume even though the resource is defined as "binds".

**Comment by chasebolt** [comment]
Good debugging! I think this may be an issue with the gem docker-api. The docker-api gem specifies API spec 1.16, volume and bind changes in a later version of the API.

If you have time to keep hacking away, change this line to 1.21
https://github.com/someara/chef-docker/blob/master/files/default/vendor/gems/docker-api-1.22.4/lib/docker/version.rb#L6

Get another dump of the api calls and see we if it improves.

**Comment by lmickh** [comment]
No dice on just changing the api_version in the vendor/gems folder.

```
{"Image":"amazon/amazon-ecs-agent:v1.6.0","Labels":{},"Cmd":null,"AttachStderr":true,"AttachStdin":false,"AttachStdout":true,"Domainname":"","Entrypoint":["/agent"],"Env":["ECS_LOGFILE=/log/ecs-agent.log","ECS_LOGLEVEL=info","ECS_DATADIR=/data","ECS_CLUSTER=default","ECS_RESERVED_MEMORY=1024"],"ExposedPorts":{"51678/tcp":{}},"Hostname":"4f85c003f470","MacAddress":"","NetworkDisabled":false,"OpenStdin":false,"StdinOnce":false,"Tty":false,"User":"","Volumes":{"/sys/fs/cgroup:/sys/fs/cgroup:ro":{},"/var/lib/ecs/data:/data":{},"/var/log/ecs:/log":{},"/var/run/docker.sock:/var/run/docker.sock":{},"/var/run/docker/execdriver/native:/var/lib/docker/execdriver/native:ro":{}},"WorkingDir":"","HostConfig":{"Binds":["/var/log/ecs:/log","/var/run/docker.sock:/var/run/docker.sock","/var/lib/ecs/data:/data","/sys/fs/cgroup:/sys/fs/cgroup","/var/run/docker/execdriver/native:/var/lib/docker/execdriver/native"],"CapAdd":null,"CapDrop":null,"CgroupParent":"","CpuShares":0,"CpusetCpus":"","Devices":null,"Dns":null,"DnsSearch":null,"ExtraHosts":null,"Links":null,"LogConfig":null,"Memory":0,"MemorySwap":-1,"NetworkMode":"default","Privileged":false,"PortBindings":{"51678/tcp":[{"HostIp":"127.0.0.1","HostPort":"51678"}]},"PublishAllPorts":false,"RestartPolicy":{"Name":"on-failure","MaximumRetryCount":10},"Ulimits":null,"VolumesFrom":null}}
```

This is looking like an upstream issue.

**Comment by chasebolt** [comment]
I am interested in knowing how Volumes is being passed. Glancing over the cookbook it should be nil. Will need to do some dumping of the `Docker::Container.create()` options and also testing a basic curl call to the api; verify the docker-api gem isn't doing something with the Volumes.

**Comment by lmickh** [comment]
So it turns out this is not from behavior of the docker-api gem.  I rebuilt the machine from scratch and it worked fine.  It appears the volume attr for the container resource are not getting cleared when I switched from volume to binds the first time.  I thought I tested for this, but now I'm not sure.  I can't recreate the issue when the server never had a container resource with volumes listed.

I don't see anything in the chef code that would cause the resource values to care over after they have been removed though...
